File: weird_captcha_gym/shared_runtime/server/incubator_graders/borrowed_tiles.py
from __future__ import annotations

from typing import Any


MECHANIC_ID = "borrowed_tiles"


def _fail(message: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "feedback": message}


def _binding_error(payload: dict[str, Any], truth: dict[str, Any], public: dict[str, Any]) -> str | None:
    if payload.get("mechanic_id") != MECHANIC_ID or truth.get("mechanic_id") != MECHANIC_ID or public.get("mechanic_id") != MECHANIC_ID:
        return "mechanic mismatch"
    challenge_id = str(truth.get("challenge_id") or "")
    if not challenge_id or payload.get("challenge_id") != challenge_id:
        return "stale challenge"
    if public.get("challenge_id") != challenge_id:
        return "public-state challenge mismatch"
    task_id = str(truth.get("task_id") or "")
    if not task_id or payload.get("task_id") != task_id or public.get("task_id") != task_id:
        return "task mismatch"
    return None


def _tile_catalog(value: Any) -> dict[str, dict[str, Any]]:
    if not isinstance(value, list) or not value:
        raise ValueError("tile catalogue is missing")
    result: dict[str, dict[str, Any]] = {}
    for tile in value:
        if not isinstance(tile, dict):
            raise ValueError("tile catalogue item is malformed")
        tile_id = str(tile.get("id") or "")
        color = tile.get("color")
        number = tile.get("number")
        if not tile_id or tile_id in result or not isinstance(color, str) or isinstance(number, bool) or not isinstance(number, int):
            raise ValueError("tile catalogue item is invalid")
        result[tile_id] = {"id": tile_id, "color": color, "number": number}
    return result


def _initial_locations(truth: dict[str, Any], catalog: dict[str, dict[str, Any]]) -> tuple[dict[str, dict[str, str]], dict[str, list[str]]]:
    sets = truth.get("initial_sets")
    rack = truth.get("initial_rack")
    if not isinstance(sets, list) or not isinstance(rack, list):
        raise ValueError("initial tile layout is missing")
    locations: dict[str, dict[str, str]] = {}
    layout: dict[str, list[str]] = {}
    for item in sets:
        if not isinstance(item, dict) or not isinstance(item.get("id"), str) or not isinstance(item.get("tiles"), list):
            raise ValueError("initial set is malformed")
        set_id = item["id"]
        if set_id in layout:
            raise ValueError("duplicate initial set")
        layout[set_id] = []
        for tile_id in item["tiles"]:
            tile_id = str(tile_id)
            if tile_id not in catalog or tile_id in locations:
                raise ValueError("initial layout does not conserve tiles")
            layout[set_id].append(tile_id)
            locations[tile_id] = {"zone": "table", "set_id": set_id}
    for tile_id in rack:
        tile_id = str(tile_id)
        if tile_id not in catalog or tile_id in locations:
            raise ValueError("initial rack does not conserve tiles")
        locations[tile_id] = {"zone": "rack"}
    if set(locations) != set(catalog):
        raise ValueError("initial layout omits or invents a tile")
    return locations, layout


def _is_valid_set(tile_ids: list[str], catalog: dict[str, dict[str, Any]]) -> tuple[bool, str]:
    if len(tile_ids) < 3:
        return False, "too short"
    tiles = [catalog[tile_id] for tile_id in tile_ids]
    numbers = [int(tile["number"]) for tile in tiles]
    colors = [str(tile["color"]) for tile in tiles]
    if len(set(numbers)) == 1 and len(tile_ids) <= 4 and len(set(colors)) == len(colors):
        return True, "group"
    if len(set(colors)) == 1 and len(set(numbers)) == len(numbers) and sorted(numbers) == list(range(min(numbers), max(numbers) + 1)):
        return True, "run"
    return False, "not a legal run or group"


def _copy_layout(initial_layout: dict[str, list[str]], initial_rack: list[str]) -> tuple[dict[str, list[str]], list[str]]:
    return {set_id: tiles[:] for set_id, tiles in initial_layout.items()}, [str(tile_id) for tile_id in initial_rack]


def _location(layout: dict[str, list[str]], rack: list[str], tile_id: str) -> dict[str, str] | None:
    if tile_id in rack:
        return {"zone": "rack"}
    for set_id, tiles in layout.items():
        if tile_id in tiles:
            return {"zone": "table", "set_id": set_id}
    return None


def _remove(layout: dict[str, list[str]], rack: list[str], tile_id: str) -> None:
    if tile_id in rack:
        rack.remove(tile_id)
        return
    for tiles in layout.values():
        if tile_id in tiles:
            tiles.remove(tile_id)
            return
    raise ValueError("tile is not in the current layout")


def _normal_layout(value: Any) -> dict[str, list[str]]:
    if not isinstance(value, list):
        raise ValueError("table layout must be a list")
    result: dict[str, list[str]] = {}
    for item in value:
        if not isinstance(item, dict) or not isinstance(item.get("id"), str) or not isinstance(item.get("tiles"), list):
            raise ValueError("table set is malformed")
        set_id = item["id"]
        if set_id in result or any(not isinstance(tile_id, str) for tile_id in item["tiles"]):
            raise ValueError("table set identity is invalid")
        result[set_id] = list(item["tiles"])
    return result


def grade(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    error = _binding_error(payload, ground_truth, public_state)
    if error:
        return _fail(error)
    if ground_truth.get("control_condition") != public_state.get("control_condition"):
        return _fail("interaction condition differs from tile contract")
    condition = ground_truth.get("control_condition") or {}
    expected_source = {"simplified": "tile_click_drop", "full": "tile_drag"}.get(str(condition.get("interaction") or "")) if condition else "tile_drag"
    if expected_source is None:
        return _fail("invalid tile interaction condition")
    try:
        catalog = _tile_catalog(ground_truth.get("tiles"))
        public_catalog = _tile_catalog(public_state.get("tiles"))
        if catalog != public_catalog:
            return _fail("public tile catalogue differs from hidden contract")
        initial_locations, initial_layout = _initial_locations(ground_truth, catalog)
        public_initial = _normal_layout(public_state.get("table_sets"))
        if {key: sorted(value) for key, value in public_initial.items()} != {key: sorted(value) for key, value in initial_layout.items()}:
            return _fail("public table layout differs from hidden contract")
        initial_rack = [str(tile_id) for tile_id in ground_truth.get("initial_rack") or []]
        if public_state.get("rack_tiles") != initial_rack:
            return _fail("public rack differs from hidden contract")
    except (TypeError, ValueError) as exc:
        return _fail(f"invalid tile contract: {exc}")

    layout, rack = _copy_layout(initial_layout, initial_rack)
    actions = payload.get("actions")
    if not isinstance(actions, list) or not (1 <= len(actions) <= 300):
        return _fail("move transcript is missing or outside limits")
    reassigned = 0
    rack_to_table = 0
    for index, action in enumerate(actions, start=1):
        if not isinstance(action, dict) or action.get("sequence") != index or action.get("input_source") != expected_source:
            return _fail(f"move {index} has invalid sequence or interaction source")
        tile_id = action.get("tile_id")
        if tile_id not in catalog:
            return _fail(f"move {index} names an unknown tile")
        actual_from = _location(layout, rack, tile_id)
        declared_from = action.get("from")
        destination = action.get("to")
        if actual_from != declared_from:
            return _fail(f"move {index} starts from the wrong visible location")
        if not isinstance(destination, dict) or destination.get("zone") not in {"rack", "table"}:
            return _fail(f"move {index} has an invalid destination")
        from_set = str(actual_from.get("set_id") or "") if actual_from.get("zone") == "table" else ""
        if destination.get("zone") == "rack":
            if actual_from.get("zone") == "rack":
                return _fail(f"move {index} is a no-op")
            _remove(layout, rack, tile_id)
            rack.append(tile_id)
        else:
            set_id = str(destination.get("set_id") or "")
            if not set_id:
                return _fail(f"move {index} has no target set")
            if set_id not in layout:
                if not set_id.startswith("new_set_") or set_id in layout:
                    return _fail(f"move {index} invents an invalid set identity")
                layout[set_id] = []
            if actual_from.get("zone") == "table" and from_set == set_id:
                return _fail(f"move {index} is a no-op")
            _remove(layout, rack, tile_id)
            layout[set_id].append(tile_id)
            if actual_from.get("zone") == "rack":
                rack_to_table += 1
            elif from_set != set_id:
                reassigned += 1

    try:
        final_layout = _normal_layout(payload.get("table_sets"))
        final_rack = payload.get("rack_tiles")
        if not isinstance(final_rack, list) or any(not isinstance(tile_id, str) for tile_id in final_rack):
            return _fail("submitted rack is malformed")
    except ValueError as exc:
        return _fail(str(exc))
    if {key: sorted(value) for key, value in final_layout.items()} != {key: sorted(value) for key, value in layout.items()} or final_rack != rack:
        return _fail("submitted layout does not match the physical move transcript")
    all_ids = [tile_id for tiles in final_layout.values() for tile_id in tiles] + list(final_rack)
    if len(all_ids) != len(catalog) or set(all_ids) != set(catalog) or len(set(all_ids)) != len(all_ids):
        return _fail("final layout does not conserve each physical tile exactly once")
    invalid_sets: list[str] = []
    valid_kinds: dict[str, str] = {}
    for set_id, tile_ids in final_layout.items():
        valid, kind = _is_valid_set(tile_ids, catalog)
        if not valid:
            invalid_sets.append(set_id)
        else:
            valid_kinds[set_id] = kind
    required_borrow = int(ground_truth.get("required_borrow_count") or 1)
    passed = (
        payload.get("completed") is True
        and not final_rack
        and not invalid_sets
        and rack_to_table >= required_borrow
        and reassigned >= required_borrow
    )
    return {
        "graded": True,
        "passed": passed,
        "feedback": (
            f"{len(final_layout)} valid sets; rack {len(final_rack)}; "
            f"rack-to-table {rack_to_table}/{required_borrow}; borrowed reassignments {reassigned}/{required_borrow}; "
            f"kinds {', '.join(valid_kinds.values()) or 'none'}"
            + (f"; invalid {', '.join(invalid_sets)}" if invalid_sets else "")
        ),
    }


def cheat(public_state: dict[str, Any], ground_truth: dict[str, Any]) -> dict[str, Any]:
    actions = []
    condition = ground_truth.get("control_condition") or {}
    source = {"simplified": "tile_click_drop", "full": "tile_drag"}.get(str(condition.get("interaction") or "full"), "tile_drag")
    for item in ground_truth.get("solution_actions") or []:
        action = dict(item)
        action["input_source"] = source
        actions.append(action)
    return {"actions": actions, "instruction": "Follow the visible ceramic tiles, borrow every required donor, then commit the table."}

File: weird_captcha_gym/shared_runtime/app/mechanics/borrowed_tiles.js
(() => {
  "use strict";

  const model = {
    state: null,
    tileMap: {},
    sets: [],
    rack: [],
    actions: [],
    selectedTile: null,
    pointerDrag: null,
    busy: false,
    terminal: false,
    newSetCounter: 0,
    helpers: null,
  };

  const clean = (value) => String(value == null ? "" : value)
    .replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;");
  const clone = (value) => JSON.parse(JSON.stringify(value));
  const tile = (id) => model.tileMap[id];

  function currentLocation(tileId) {
    if (model.rack.includes(tileId)) return {zone: "rack"};
    const group = model.sets.find((entry) => entry.tiles.includes(tileId));
    return group ? {zone: "table", set_id: group.id} : null;
  }

  function validKind(tileIds) {
    if (tileIds.length < 3) return "needs 3+ tiles";
    const values = tileIds.map(tile);
    const numbers = values.map((entry) => Number(entry.number));
    const colors = values.map((entry) => String(entry.color));
    if (new Set(numbers).size === 1 && tileIds.length <= 4 && new Set(colors).size === colors.length) return "VALID GROUP";
    const ordered = [...numbers].sort((a, b) => a - b);
    if (new Set(colors).size === 1 && new Set(numbers).size === numbers.length && ordered.every((value, index) => value === ordered[0] + index)) return "VALID RUN";
    return "NEEDS REPAIR";
  }

  function isValid(tileIds) {
    return validKind(tileIds).startsWith("VALID");
  }

  function tileMarkup(tileId) {
    const entry = tile(tileId);
    if (!entry) return "";
    const label = `${entry.color} ${entry.number}`;
    return `<button type="button" class="borrowed-tile glaze-${Number(entry.pattern || 0)}" data-tile-id="${clean(tileId)}" aria-label="${clean(label)}">
      <span class="tile-corner">${Number(entry.number)}</span><strong>${Number(entry.number)}</strong><small>${clean(entry.color)}</small>
    </button>`;
  }

  function setMarkup(entry) {
    const kind = validKind(entry.tiles);
    const statusClass = kind.startsWith("VALID") ? "is-valid" : "is-open";
    return `<article class="borrow-set ${statusClass}" data-set-id="${clean(entry.id)}">
      <header><span>${clean(entry.label)}</span><b>${kind}</b></header>
      <div class="borrow-set-tiles">${entry.tiles.map(tileMarkup).join("")}</div>
      <div class="borrow-dropzone borrow-set-dropzone" data-drop-zone="table" data-set-id="${clean(entry.id)}">DROP TILE HERE</div>
    </article>`;
  }

  function rackMarkup() {
    const state = model.rack.length ? `${model.rack.length} TILE${model.rack.length === 1 ? "" : "S"} REMAIN` : "RACK EMPTY";
    return `<section class="borrow-rack-zone">
      <div class="borrow-rack-head"><span>YOUR RACK</span><b>${state}</b></div>
      <div class="borrow-rack-tiles">${model.rack.map(tileMarkup).join("")}</div>
      <div class="borrow-dropzone borrow-rack-dropzone" data-drop-zone="rack">RETURN TILE TO RACK</div>
    </section>`;
  }

  function newSetMarkup() {
    return `<div class="borrow-new-set"><div class="borrow-dropzone" data-drop-zone="new">＋ DROP TO START A NEW SET</div><small>Split a run or group here, then repair it.</small></div>`;
  }

  function renderLayout() {
    const table = document.getElementById("borrow-table");
    const rack = document.getElementById("borrow-rack-wrap");
    const newZone = document.getElementById("borrow-new-zone");
    if (table) table.innerHTML = model.sets.map(setMarkup).join("");
    if (rack) rack.innerHTML = rackMarkup();
    if (newZone) newZone.innerHTML = newSetMarkup();
    bindControls();
    updateReadout();
  }

  function updateReadout() {
    const count = model.sets.filter((entry) => isValid(entry.tiles)).length;
    const needed = model.sets.length;
    const valid = count === needed && model.rack.length === 0 && model.actions.length > 0;
    const commit = document.getElementById("borrow-commit");
    const meter = document.getElementById("borrow-valid-meter");
    const moves = document.getElementById("borrow-move-count");
    if (meter) meter.textContent = `${count} / ${needed} SETS VALID`;
    if (moves) moves.textContent = String(model.actions.length).padStart(2, "0");
    if (commit) commit.classList.toggle("is-ready", valid);
  }

  function clearVerdict() {
    document.querySelectorAll(".borrow-verdict").forEach((node) => node.remove());
    document.querySelector(".borrowed-tiles-captcha")?.classList.remove("is-fresh-fail");
  }

  function removeTile(tileId) {
    const rackIndex = model.rack.indexOf(tileId);
    if (rackIndex >= 0) {
      model.rack.splice(rackIndex, 1);
      return;
    }
    const source = model.sets.find((entry) => entry.tiles.includes(tileId));
    if (!source) throw new Error("tile is not in the visible layout");
    source.tiles.splice(source.tiles.indexOf(tileId), 1);
  }

  function ensureNewSet() {
    let id;
    do {
      model.newSetCounter += 1;
      id = `new_set_${model.newSetCounter}`;
    } while (model.sets.some((entry) => entry.id === id));
    model.sets.push({id, label: `FRESH SET ${model.newSetCounter}`, tiles: []});
    return id;
  }

  function moveTile(tileId, destination, inputSource) {
    if (model.busy || model.terminal) return;
    const from = currentLocation(tileId);
    if (!from) return;
    const target = destination.zone === "new" ? {zone: "table", set_id: ensureNewSet()} : destination;
    if (target.zone === from.zone && target.set_id === from.set_id) {
      model.helpers.setReadout("CHOOSE A DIFFERENT DROP ZONE", "error");
      return;
    }
    clearVerdict();
    removeTile(tileId);
    if (target.zone === "rack") {
      model.rack.push(tileId);
    } else {
      const set = model.sets.find((entry) => entry.id === target.set_id);
      if (!set) return;
      set.tiles.push(tileId);
    }
    model.actions.push({
      sequence: model.actions.length + 1,
      tile_id: tileId,
      from: clone(from),
      to: clone(target),
      input_source: inputSource,
    });
    model.selectedTile = null;
    model.helpers.setReadout(`MOVED ${tile(tileId).color.toUpperCase()} ${tile(tileId).number} · CHECK EVERY SET`, "idle");
    renderLayout();
  }

  function selectTile(tileId) {
    if (model.busy || model.terminal) return;
    clearVerdict();
    if (model.selectedTile === tileId) {
      model.selectedTile = null;
      renderLayout();
      return;
    }
    model.selectedTile = tileId;
    model.helpers.setReadout("TILE HELD · CHOOSE A DROP ZONE", "idle");
    document.querySelectorAll(".borrowed-tile").forEach((node) => node.classList.toggle("is-selected", node.dataset.tileId === tileId));
  }

  function dragMove(event) {
    const drag = model.pointerDrag;
    if (!drag || drag.pointerId !== event.pointerId) return;
    event.preventDefault();
    drag.node.style.setProperty("--drag-x", `${event.clientX - drag.startX}px`);
    drag.node.style.setProperty("--drag-y", `${event.clientY - drag.startY}px`);
  }

  function dragEnd(event, cancelled = false) {
    const drag = model.pointerDrag;
    if (!drag || drag.pointerId !== event.pointerId) return;
    event.preventDefault();
    const dropNode = cancelled ? null : document.elementsFromPoint(event.clientX, event.clientY).find((node) => node.dataset?.dropZone);
    model.pointerDrag = null;
    drag.node.classList.remove("is-dragging");
    drag.node.style.removeProperty("--drag-x");
    drag.node.style.removeProperty("--drag-y");
    try { drag.node.releasePointerCapture(event.pointerId); } catch (_error) { /* already released */ }
    if (!dropNode) {
      model.helpers.setReadout("DROP CANCELED · TILE RETURNED", "error");
      return;
    }
    moveTile(drag.tileId, {zone: dropNode.dataset.dropZone, set_id: dropNode.dataset.setId}, "tile_drag");
  }

  function bindControls() {
    const interaction = model.state.control_condition?.interaction || "full";
    document.querySelectorAll(".borrowed-tile").forEach((node) => {
      node.addEventListener("click", (event) => {
        if (interaction === "simplified") {
          event.stopPropagation();
          selectTile(node.dataset.tileId);
        }
      });
      node.addEventListener("pointerdown", (event) => {
        if (interaction !== "full" || event.button !== 0 || model.busy || model.terminal || model.pointerDrag) return;
        event.preventDefault();
        model.pointerDrag = {pointerId: event.pointerId, tileId: node.dataset.tileId, node, startX: event.clientX, startY: event.clientY};
        node.classList.add("is-dragging");
        try { node.setPointerCapture(event.pointerId); } catch (_error) { /* no capture */ }
      });
      node.addEventListener("pointermove", dragMove);
      node.addEventListener("pointerup", (event) => dragEnd(event));
      node.addEventListener("pointercancel", (event) => dragEnd(event, true));
    });
    document.querySelectorAll("[data-drop-zone]").forEach((node) => {
      node.addEventListener("click", () => {
        if (interaction !== "simplified" || !model.selectedTile) return;
        moveTile(model.selectedTile, {zone: node.dataset.dropZone, set_id: node.dataset.setId}, "tile_click_drop");
      });
    });
  }

  async function commit() {
    if (model.busy || model.terminal) return;
    model.busy = true;
    document.querySelectorAll("button").forEach((node) => { node.disabled = true; });
    model.helpers.setReadout("AUDITING THE TABLE…", "pending");
    const payload = {
      mechanic_id: model.state.mechanic_id,
      task_id: model.state.task_id,
      challenge_id: model.state.challenge_id,
      actions: clone(model.actions),
      table_sets: model.sets.map((entry) => ({id: entry.id, tiles: entry.tiles.slice()})),
      rack_tiles: model.rack.slice(),
      completed: true,
    };
    try {
      const response = await fetch("/result", {method: "POST", headers: {"content-type": "application/json"}, body: JSON.stringify(payload)});
      const outcome = await response.json();
      if (outcome.passed === true) {
        model.terminal = true;
        document.querySelector(".borrowed-tiles-captcha")?.classList.add("is-pass");
        document.querySelector(".borrowed-tiles-captcha")?.insertAdjacentHTML("beforeend", '<div class="borrow-verdict borrow-verdict-pass"><small>TABLE BALANCED</small><strong>PASS</strong></div>');
        model.helpers.setReadout("PASS", "passed");
      } else if (outcome.passed === false && outcome.state) {
        await model.helpers.render(outcome.state);
        document.querySelector(".borrowed-tiles-captcha")?.classList.add("is-fresh-fail");
        document.querySelector(".borrowed-tiles-captcha")?.insertAdjacentHTML("beforeend", '<div class="borrow-verdict borrow-verdict-fail"><small>TABLE REJECTED · FRESH DEAL</small><strong>FAIL</strong></div>');
        model.helpers.setReadout("FAIL · FRESH DEAL", "error");
      } else {
        model.busy = false;
        document.querySelectorAll("button").forEach((node) => { node.disabled = false; });
        model.helpers.setReadout("FAIL · TABLE NOT ACCEPTED", "error");
      }
    } catch (_error) {
      model.busy = false;
      document.querySelectorAll("button").forEach((node) => { node.disabled = false; });
      model.helpers.setReadout("FAIL · REGISTER OFFLINE", "error");
    }
  }

  function reset() {
    if (model.busy || model.terminal) return;
    model.sets = clone(model.state.table_sets);
    model.rack = model.state.rack_tiles.slice();
    model.actions = [];
    model.selectedTile = null;
    model.newSetCounter = 0;
    clearVerdict();
    renderLayout();
    model.helpers.setReadout("WORKTABLE RESET · NO MOVES RECORDED", "idle");
  }

  function render(state, helpers) {
    document.body.dataset.mechanic = "borrowed-tiles";
    document.body.dataset.borrowPalette = String(state.palette || "terracotta");
    document.body.dataset.cheatMode = helpers.isCheatMode() ? "true" : "false";
    Object.assign(model, {
      state,
      tileMap: Object.fromEntries((state.tiles || []).map((entry) => [entry.id, entry])),
      sets: clone(state.table_sets || []),
      rack: (state.rack_tiles || []).slice(),
      actions: [],
      selectedTile: null,
      pointerDrag: null,
      busy: false,
      terminal: false,
      newSetCounter: 0,
      helpers,
    });
    const interaction = state.control_condition?.interaction || "full";
    helpers.app.innerHTML = `<section class="borrowed-tiles-captcha" data-interaction="${clean(interaction)}" data-challenge-id="${clean(state.challenge_id)}">
      <header class="borrow-header"><div class="borrow-title"><span>ARCHIVE OF LOOSE CERAMICS · TABLE ${String(state.generator?.module_count || 0).padStart(2, "0")}</span><h1>${clean(state.prompt)}</h1></div><div class="borrow-ticket"><small>CONSERVE EVERY IDENTITY</small><strong>${(state.tiles || []).length}</strong><i>TILES</i></div></header>
      <div class="borrow-rulebar"><span>RUN · 3+ CONSECUTIVE / ONE COLOUR</span><span>GROUP · 3–4 EQUAL / DIFFERENT COLOURS</span><b>${interaction === "full" ? "DIRECT DRAG" : "SELECT THEN DROP"}</b></div>
      <main class="borrow-workbench"><section class="borrow-table-case"><div class="borrow-section-head"><div><span>THE TABLE</span><small>Temporary breaks are allowed. Commit only when every set is green.</small></div><b id="borrow-valid-meter">0 / 0 SETS VALID</b></div><div class="borrow-table" id="borrow-table"></div><div id="borrow-new-zone"></div></section><aside class="borrow-side"><div class="borrow-ledger"><span>MOVE LEDGER</span><strong id="borrow-move-count">00</strong><small>VISIBLE TRANSFERS</small><div class="ledger-line"></div><p>Every tile is tracked by identity. A borrowed table piece must finish in a different set.</p></div><div class="borrow-legend"><span>READ THE GLAZE</span><i class="legend-coral"></i><i class="legend-azure"></i><i class="legend-amber"></i><i class="legend-jade"></i><small>Colour names are printed on each tile.</small></div></aside></main><div id="borrow-rack-wrap"></div><footer class="borrow-footer"><button type="button" id="borrow-reset">↺ RESET WORKTABLE</button><div class="readout" data-status="idle">${interaction === "full" ? "DRAG A TILE TO A SET" : "SELECT A TILE, THEN A DROP ZONE"}</div><button type="button" id="borrow-commit">COMMIT TABLE</button></footer>${helpers.cheatPanelTemplate()}</section>`;
    renderLayout();
    document.getElementById("borrow-reset")?.addEventListener("click", reset);
    document.getElementById("borrow-commit")?.addEventListener("click", commit);
    helpers.installCheatPanel();
    window.borrowedTilesModel = model;
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.borrowed_tiles = {rootSelector: ".borrowed-tiles-captcha", render};
})();
